`$onInit` and friends

petebacondarwin opened the pull request.

petebacondarwin committed:
feat($compile): call `$ngOnInit` on directive controllers after controller construction

This enables option three of https://github.com/angular/angular.js/issues/13510#issuecomment-164140194
by allowing the creator of directive controllers using ES6 classes to have a hook
that is called when the bindings are definitely available.

Moreover this will help solve the problem of accessing `require`d controllers
from controller instances without resorting to wiring up in a `link` function.
See https://github.com/angular/angular.js/issues/5893

petebacondarwin committed:
test($compile): check that $onInit is called correctly for ES6 classes

petebacondarwin committed:
docs($compile): document the new `$onInit` controller hook

petebacondarwin committed:
feat($compile): allow `require` to be an object

This provides an elegant alternative to the array form of the `require`
property but also helps to support binding of `require`d controllers
to directive controllers.

petebacondarwin committed:
feat($compile): allow required controllers to be bound to the directive controller

If directives are required through an object hash, rather than a string or array,
the required directives' controllers are bound to the current directive's controller
in much the same way as the properties are bound to using `bindToController`.

The binding is done after the controller has been constructed and all the bindings
are guaranteed to be complete by the time the controller's `$onInit` method
is called.

This change makes it much simpler to access require controllers without the
need for manually wiring them up in link functions. In particular this
enables support for `require` in directives defined using `mod.component()`

gkalpak commented on src/ng/compile.js:
```diff
@@ -255,12 +265,22 @@
+ * * a **string** containing the name of the directive to pass to the linking function
+ * * an **array** containing the names of directives to pass to the linking function. The argument passed to the
+ * linking function will be an array of controllers in the same order as the names in the `require` property
+ * * an **object** whose property values are the names of the directibes to pass to the linking function. The argument
```
directibes  -->  directives

gkalpak commented on src/ng/compile.js:
```diff
@@ -255,12 +265,22 @@
+ * linking function will be an array of controllers in the same order as the names in the `require` property
+ * * an **object** whose property values are the names of the directibes to pass to the linking function. The argument
+ * passed to the linking function will also be an object with matching keys, whose values will be the corresponding
+ * controller.
```
controller  -->  controllers (?)

gkalpak commented on src/ng/compile.js:
```diff
@@ -2386,8 +2412,20 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+
+          if (isObject(controllerDirective.require) && !isArray(controllerDirective.require)) {
+            var controllerObj = getControllers(name, controllerDirective.require, $element, elementControllers);
+            extend(controller.instance, controllerObj);
```
Doesn't this mean that controllers whose constructors return a value won't be properly handled ?
Do we really need to set the required controllers as properties on the requiring controller ? How about just passing the object as argument to `$onInit` ?

petebacondarwin replied:
By the time we get here we have already dealt with constructors who return a value (around lines 2406-2414). I have added a test to prove this:

petebacondarwin replied:
I didn't like to add parameters to `$onInit`. If we don't want to automatically bind then, instead, we should be injecting things into the constructor of the controller, like in ng2... But that is done differently for non-ancestor components

gkalpak replied:
I meant if one of the **required** controllers returns a value, then we might extend **this** controller with the wrong value (depends on the order of execution of the directives).

(Your test proves that it works if **this** controller returns a value.)

petebacondarwin replied:
You are right! I have fixed it and added another test in 1d18df20fe6dd0ab50f9285f03ff979f302f5874

gkalpak replied:
LGTM

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -9435,5 +9530,5 @@ describe('$compile', function() {
       });
     });
-  });
+   });
```
I am on mobile (so I can't be sure), but tjis change doesn't seem right...

gkalpak commented:
With the exception of https://github.com/angular/angular.js/pull/13763#discussion_r49637240 (and a couple of nitpicks), this L(pretty)GTM :+1:

petebacondarwin committed:
feat($compile): allow `require` to be an object

This provides an elegant alternative to the array form of the `require`
property but also helps to support binding of `require`d controllers
to directive controllers.

petebacondarwin committed:
test($compile): check explicit return controllers are not broken by binding require

petebacondarwin committed:
feat($compile): allow required controllers to be bound to the directive controller

If directives are required through an object hash, rather than a string or array,
the required directives' controllers are bound to the current directive's controller
in much the same way as the properties are bound to using `bindToController`.

The binding is done after the controller has been constructed and all the bindings
are guaranteed to be complete by the time the controller's `$onInit` method
is called.

This change makes it much simpler to access require controllers without the
need for manually wiring them up in link functions. In particular this
enables support for `require` in directives defined using `mod.component()`

petebacondarwin committed:
feat($compile): call `$ngOnInit` on directive controllers after controller construction

This enables option three of https://github.com/angular/angular.js/issues/13510#issuecomment-164140194
by allowing the creator of directive controllers using ES6 classes to have a hook
that is called when the bindings are definitely available.

Moreover this will help solve the problem of accessing `require`d controllers
from controller instances without resorting to wiring up in a `link` function.
See https://github.com/angular/angular.js/issues/5893

petebacondarwin committed:
feat($compile): allow required controllers to be bound to the directive controller

If directives are required through an object hash, rather than a string or array,
the required directives' controllers are bound to the current directive's controller
in much the same way as the properties are bound to using `bindToController`.

The binding is done after the controller has been constructed and all the bindings
are guaranteed to be complete by the time the controller's `$onInit` method
is called.

This change makes it much simpler to access require controllers without the
need for manually wiring them up in link functions. In particular this
enables support for `require` in directives defined using `mod.component()`

lgalfaso commented on test/ng/compileSpec.js:
```diff
@@ -4244,6 +4244,22 @@ describe('$compile', function() {
       if (!/chrome/i.test(navigator.userAgent)) return;
```
unrelated to this issue, but can we change this to `if (!support.classes) return;` ?

petebacondarwin replied:
On it [sic]!

petebacondarwin replied:
Actually, it is not so easy @lgalfaso since the latest Safari (9.0.2) is even worse for `toString` on a a class instance.

lgalfaso replied:
it looks like there are at least 2 issues reported on JavascriptCore (Safari JS engine) regarding this
https://bugs.webkit.org/show_bug.cgi?id=149743
https://bugs.webkit.org/show_bug.cgi?id=144285

I think that some other strategy. Meanwhile, just keep the old code

Narretz commented on src/ng/compile.js:
```diff
@@ -255,12 +265,26 @@
+ *
+ * If the `require` property is an object and the directive provides a controller, then the required controllers are
+ * bound to the controller using the keys of the `require` property. See the {@link $compileProvider#component} helper
+ * for an example of how this can be used.
```
Should note that the controllers are only available once $onInit has been called, no?

petebacondarwin replied:
Updated

Narretz commented on src/ng/compile.js:
```diff
@@ -255,12 +265,26 @@
  *    The `$transclude` function also has a method on it, `$transclude.isSlotFilled(slotName)`, which returns
  *    `true` if the specified slot contains content (i.e. one or more DOM nodes).
  *
+ * The controller can provide the following methods that act as life-cycle hooks:
```
Should we mention that this is basically a new phase that happens before both pre and postLink?

petebacondarwin replied:
Updated

Narretz commented:
Now that require can be an object, should we pass the ctrls as an object to the link function like described here: https://github.com/angular/angular.js/pull/8401?

Narretz commented:
Don't forget that these issues are fixed by the PR:
https://github.com/angular/angular.js/pull/6040
https://github.com/angular/angular.js/issues/5893

petebacondarwin commented:
> Now that require can be an object, should we pass the ctrls as an object to the link function like described here: #8401? (Yes from me)

@Narretz One of the commits in this PR does actually do that - but only when the `require` property is an object itself.

petebacondarwin committed:
fix($compile): ensure controllers with return value constructors are required correctly

See https://github.com/angular/angular.js/pull/13763#discussion_r49836041

petebacondarwin committed:
docs($compile): squash me

petebacondarwin committed:
fix($compile): only bind required controllers if `bindToController` is truthy

Narretz commented on src/ng/compile.js:
```diff
@@ -2388,6 +2494,21 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
           }
         }
 
+        // Bind the required controllers to the controller, if `require` is an object
```
... and bindToController is truthy

gkalpak replied:
Nitpick: Comment could also mention "bindToController" as a requirement.

petebacondarwin replied:
done

gkalpak commented on src/ng/compile.js:
```diff
@@ -218,8 +218,18 @@
- * is instantiated, the initial values of the isolate scope bindings will be available if the controller is not an ES6 class.
+ * allow a component to have its properties bound to the controller, rather than to scope.
+ *
+ * After the controller is instantiated, the initial values of the isolate scope bindings will bound to the controller
```
will bound  -->  will be? bound

petebacondarwin replied:
done

gkalpak commented on src/ng/compile.js:
```diff
@@ -255,12 +265,28 @@
+ * See the {@link $compileProvider#component} helper for an example of how this can be used.
+ *
+ * If no such required directive(s) can be found, or if the directive does not have a controller, then an error is
+ * raised (unless no link function is specified, in which case error checking is skipped). The name can be prefixed with:
```
Is error checking indeed skipped if no link function is specified ?

petebacondarwin replied:
indeed - the error checking is inside `getControllers` which is not called if no link function (and now also no bindToController+require-as-object)

gkalpak replied:
I meant don't we need to mention that error-checking is skipped if no link function **AND** no bindToController+requireAsObject ?

petebacondarwin replied:
fixed

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4280,6 +4283,7 @@ describe('$compile', function() {
                                  'dir-str="Hello, {{whom}}!" ' +
                                  'dir-fn="fn()"></div>')($rootScope);
+        expect(Controller.prototype.$onInit).toHaveBeenCalled();
         element.data('$fooDirController').check();
```
Isn't this redundant now ?

petebacondarwin replied:
yep! removed

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4287,6 +4291,7 @@ describe('$compile', function() {
     });
 
 
+
```
Too much whitespace ?

petebacondarwin replied:
removed

Narretz commented:
I just noticed the component method doesn't accept the templateNamespace property, is this intentional?

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -5307,6 +5338,195 @@ describe('$compile', function() {
+            require: { container: '^parent', friend: 'sibling' },
+            bindToController: true,
+            controller: MeController,
+            controllerAs: '$ctrl'
```
Two `controllerAs` occurrences (where 0 would be sufficient ?).

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -5307,6 +5338,195 @@ describe('$compile', function() {
+    });
+
+
+    it('should not bind required controllers bindToController is falsy', function() {
```
Missing "if".

petebacondarwin replied:
fixed

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -5307,6 +5338,195 @@ describe('$compile', function() {
+    });
+
+
+    it('should bind required controllers to controllers that return an explicit constructor return value', function() {
```
Nitpick: I'd add a priority to `me` (1 would do), in order to make the test more explicit (and future-proof).

Explanation:
Currently, this test relies on the fact that the directives have default priorities, which means that the order of execution of directives on the same element is determined by their alphabetical order. Thus `me` gets a "higher
 priority than `sibling` and its controller is instantiated first.
<sub>(I said it's a nitpick :stuck_out_tongue: )</sub>

petebacondarwin replied:
OK, fair enough

petebacondarwin replied:
Done

petebacondarwin committed:
docs($compile): fix typo

petebacondarwin committed:
fix($compile): only bind required controllers if `bindToController` is truthy

petebacondarwin committed:
test($compile): check that $onInit is called correctly for ES6 classes

petebacondarwin committed:
fix($compile): only bind required controllers if `bindToController` is truthy

petebacondarwin committed:
fix($compile): ensure controllers with return value constructors are required correctly

See https://github.com/angular/angular.js/pull/13763#discussion_r49836041

petebacondarwin committed:
fix($compile): only bind required controllers if `bindToController` is truthy

petebacondarwin commented:
> I just noticed the component method doesn't accept the templateNamespace / multiElement property, is this intentional?

@Narretz - I think we should try to keep the component API as simple as possible. These two are fairly uncommon, right? If people really need them they can complain and we can add them later.

Narretz commented:
@petebacondarwin Sounds good

Narretz commented:
There's a real travis failure btw:
> Chrome 45.0.2454 (Linux 0.0.0) $compile isolated locals should eventually expose isolate scope variables on ES6 class controller with controllerAs when bindToController is true FAILED

Although the test wasn't introduced by this PR.

petebacondarwin committed:
test($compile): check that $onInit is called correctly for ES6 classes

petebacondarwin commented:
@Narretz I have fixed that failing test.

ev45ive commented:
This looks very nice and clean. Should be like this from beggining. Very nice work. Do you think $onInit with bound required controllers and component() syntax will get on board with 1.5 release? I'd love to use it in production - it would clean up a lot of messy directive code. :-)
